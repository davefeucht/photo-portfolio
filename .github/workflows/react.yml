# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
