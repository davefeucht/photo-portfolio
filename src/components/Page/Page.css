.page {
  min-height: 100%;
  display: grid;
  grid-template-columns: 100%;
  grid-template-areas: "page-header" "page-content";
  grid-template-rows: .25fr 4fr;
}