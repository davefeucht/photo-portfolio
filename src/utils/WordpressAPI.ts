import axios from 'axios';

import {
    API,
    CategoriesResponse,
    CategoryResponse,
    MediaResponse,
    PageResponse,
    PagesResponse,
    PostResponse,
    PostsResponse,
    SiteInfoResponse,
    TagNameResponse
} from './types';

class WordpressAPI implements API {
    siteUrl: string;

    constructor(siteUrl: string) {
        this.siteUrl = siteUrl;
    }

    getPostThumbnail = async (featuredImage: number) => {
        const getPostThumbnailURI = `https://${this.siteUrl}/wp-json/wp/v2/media/${featuredImage}`;
        const response = await axios.get(getPostThumbnailURI);
        const size = response.data.media_details.sizes.large ? 'large' : 'full';
        const thumbUrl: string = response.data.media_details.sizes[size].source_url;

        return thumbUrl;
    };

    getSiteInfo = async () => {
        const getSiteInformationURI = `https://${this.siteUrl}/wp-json/`;

        const response: SiteInfoResponse = await axios.get(getSiteInformationURI);
        const siteName = response.data.name;

        return siteName;
    };

    getCategories = async () => {
        const getCategoriesURI = `https://${this.siteUrl}/wp-json/wp/v2/categories?exclude=175`;

        const response: CategoriesResponse = await axios.get(getCategoriesURI);
        const categories = response.data;

        return categories;
    };

    getCategoryImage = async (categoryId: number) => {
        // Define the request string to get the posts for this category
        const getCategoryPostURI = `https://${this.siteUrl}/wp-json/wp/v2/posts?categories=${categoryId}`;

        // Var to return
        let fullImageUrl = '';

        // Make the request
        const response = await axios.get(getCategoryPostURI);
        const categories = response.data;

        // Determine a random post from the ones returned
        const numberOfPosts = categories.length;
        const randomPost = Math.floor((Math.random() * numberOfPosts));

        // If this post does exist in the returned results
        if (categories[randomPost] !== undefined) {
            // Define the request string to get the featured media for the random post
            const categoryImageURI = `https://${this.siteUrl}/wp-json/wp/v2/media/${categories[randomPost].featured_media}/`;

            const imageURIResponse = await axios.get(categoryImageURI);
            const categoryImage = imageURIResponse.data;

            // If the large size image exists, use it
            if (categoryImage.media_details.sizes.large) {
                fullImageUrl = categoryImage.media_details.sizes.large.source_url;
            // Otherwise use the medium size version
            } else {
                fullImageUrl = categoryImage.media_details.sizes.medium.source_url;
            }
        }

        return fullImageUrl;
    };

    getPosts = async (categoryId: number) => {
        const getPostsURI = `https://${this.siteUrl}/wp-json/wp/v2/posts?categories=${categoryId}&per_page=20`;
        const response: PostsResponse = await axios.get(getPostsURI);
        const posts = response.data;

        return posts;
    };

    getPost = async (postId: number) => {
        const getPostURI = `https://${this.siteUrl}/wp-json/wp/v2/posts/${postId}`;
        const response: PostResponse = await axios.get(getPostURI);
        const post = response.data;

        return post;
    };

    getCategoryInfo = async (categoryId: number) => {
        const getPostsCategoryURI = `https://${this.siteUrl}/wp-json/wp/v2/categories/${categoryId}`;
        const response: CategoryResponse = await axios.get(getPostsCategoryURI);
        const categoryData = response.data;

        return categoryData;
    };

    getPostImage = async (image: number) => {
        const getPostImageURI = `https://${this.siteUrl}/wp-json/wp/v2/media/${image}`;
        const response: MediaResponse = await axios.get(getPostImageURI);
        const fullImageUrl = response.data.media_details.sizes.full?.source_url ?? '';

        return fullImageUrl;
    };

    getTagNames = async (tags: number[]) => {
        const tagNames: string[] = [];
        const tagPromises = tags.map(async tagId => {
            return axios.get(`https://${this.siteUrl}/wp-json/wp/v2/tags/${tagId}`)
                .then((response: TagNameResponse) => {
                    tagNames.push(response.data.name);
                });
        });

        await Promise.all(tagPromises);

        return tagNames;
    };

    getPages = async () => {
        const getPagesURI = `https://${this.siteUrl}/wp-json/wp/v2/pages`;
        const response: PagesResponse = await axios.get(getPagesURI);
        const pages = response.data;

        return pages;
    };

    getPage = async (pageId: number) => {
        const getPagesURI = `https://${this.siteUrl}/wp-json/wp/v2/pages/${pageId}`;
        const response: PageResponse = await axios.get(getPagesURI);
        const page = response.data;

        return page;
    };
}

export default WordpressAPI;
