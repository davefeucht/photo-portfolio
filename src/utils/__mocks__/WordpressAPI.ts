const imageUrl = 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg';

const siteInfo = {
    siteName: 'Through a Pinhole',
    siteUrl: 'throughapinhole.com'
};

const categories = [
    {
        id: 23,
        count: 6,
        description: 'Category of other things',
        link: '',
        name: 'Some other category',
        slug: 'smocg',
        taxonomy: '',
        parent: 12,
        meta: {}
    },
    {
        id: 35,
        count: 3,
        description: 'Category of things',
        link: '',
        name: 'Some category',
        slug: 'smcg',
        taxonomy: '',
        parent: 23,
        meta: {}
    }
];

const posts = [
    {
        id: 887,
        link: '',
        modified: '',
        slug: '',
        type: '',
        featured_media: 142,
        categories: [220],
        title: {
            rendered: 'Some post'
        },
        content: {
            rendered: 'Some content',
            protected: false
        },
        thumbnail_image: 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'
    },
    {
        id: 889,
        featured_media: 145,
        categories: [220],
        title: {
            rendered: 'Some other post'
        },
        thumbnail_image: 'https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg'
    }
];

const pages = [
    {
        id: 150,
        content: {
            rendered: 'Contact me!'
        },
        title: {
            rendered: 'Contact'
        }
    }
];

const tagNames = ['something', 'somethingelse'];

export default class WordpressAPI {
    siteUrl: string;

    constructor(siteUrl: string) {
        this.siteUrl = siteUrl;
    }

    async getPostThumbnail() {
        return Promise.resolve(imageUrl);
    }

    async getSiteInfo() {
        return Promise.resolve(siteInfo.siteName);
    }

    async getCategories() {
        return Promise.resolve(categories);
    }

    async getCategoryImage() {
        return Promise.resolve(imageUrl);
    }

    async getPosts() {
        return Promise.resolve(posts);
    }

    async getPost() {
        return Promise.resolve(posts[0]);
    }

    async getCategoryInfo() {
        return Promise.resolve(categories[0]);
    }

    async getPostImage() {
        return Promise.resolve(imageUrl);
    }

    async getTagNames() {
        return Promise.resolve(tagNames);
    }

    async getPages() {
        return Promise.resolve(pages);
    }

    async getPage() {
        return Promise.resolve(pages[0]);
    }
}
