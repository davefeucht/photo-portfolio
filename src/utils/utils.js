export default class Utils {
  constructor() {
    this.instance = null;
  }

  static getInstance() {
    if (this.instance) {
      return this.instance;
    } else {
      this.instance = new Utils();
      return this.instance;
    }
  }

  getSiteInfo() {
    const getSiteInformationURI = `https://${stateStore.siteInfo.siteUrl}/wp-json/`;

    axios.get(getSiteInformationURI)
      .then((response) => {
        runInAction(() => {
          stateStore.setSiteName(response.data.name);
        })
      })
      .catch(error => {
        console.warn(error.message);
      });
  }
}