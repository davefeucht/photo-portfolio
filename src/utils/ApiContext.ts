import { createContext } from 'react';

import { API } from './types';

export const ApiContext = createContext<API | null>(null);
