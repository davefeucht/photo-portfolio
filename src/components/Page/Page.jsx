/** ***************
* Page component implements displaying a page in the application
**************** */

import './Page.css';

import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';

import { getPage } from '../../utils/Api.js';
import PageContent from '../PageContent/PageContent.jsx';
import SectionHeader from '../SectionHeader/SectionHeader.jsx';

function Page({ stateStore }) {
    const { pageId } = useParams();
    const { currentPageData } = stateStore;

    useEffect(() => {
        getPage(pageId, stateStore.siteInfo.siteUrl)
            .then(page => {
                stateStore.setPageData(page);
            });
    }, [pageId]);

    return (
        <div className="page">
            <SectionHeader title={currentPageData.title?.rendered} />
            <PageContent content={currentPageData.content?.rendered} />
        </div>
    );
}

Page.displayName = 'Page';

Page.propTypes = {
    stateStore: PropTypes.object.isRequired
};

export default observer(Page);
