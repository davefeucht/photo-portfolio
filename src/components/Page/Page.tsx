/** ***************
* Page component implements displaying a page in the application
**************** */

import './Page.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect } from 'react';
import { useParams } from 'react-router-dom';

import { getPage } from '../../utils/Api';
import {
    Page as PageState,
    SiteInfo
} from '../../utils/types';
import PageContent from '../PageContent/PageContent';
import SectionHeader from '../SectionHeader/SectionHeader';

interface PageProps {
    currentPageData: PageState,
    siteInfo: SiteInfo,
    setPageData: (a: PageState) => void
}

const Page: React.FC<PageProps> = ({ currentPageData, siteInfo, setPageData }) => {
    const { pageId } = useParams();

    useEffect(() => {
        if (pageId) {
            getPage(parseInt(pageId), siteInfo.siteUrl)
                .then(page => {
                    setPageData(page);
                });
        }
    }, [pageId]);

    return (
        <div className="page">
            <SectionHeader title={currentPageData.title?.rendered} />
            <PageContent content={currentPageData.content?.rendered} />
        </div>
    );
};

Page.displayName = 'Page';

export default observer(Page);
