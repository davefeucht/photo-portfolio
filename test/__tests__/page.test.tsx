import { render } from '@testing-library/react';
import Page from 'components/Page/Page';
import * as React from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import StateStore from 'StateStore/store';
import { ApiContext } from 'utils/ApiContext';
import WordpressAPI from 'utils/WordpressAPI';

jest.mock('../../utils/WordpressAPI');

test('Page displays', async () => {
    const store = new StateStore();
    const api = new WordpressAPI(store.siteInfo.siteUrl);
    const page = await api.getPage(150);
    const { container } = render(
        <MemoryRouter initialEntries={['/page/150']}>
            <ApiContext.Provider value={api}>
                <Routes>
                    <Route
                        path="page/:pageId"
                        element={(
                            <Page
                                currentPageData={page}
                                setPageData={store.setPageData}
                            />
                        )}
                    />
                </Routes>
            </ApiContext.Provider>
        </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
});
