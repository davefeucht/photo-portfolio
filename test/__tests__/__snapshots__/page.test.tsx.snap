// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page displays 1`] = `
<div
  class="page"
>
  <div
    class="section-header"
  >
    <div
      class="title"
    >
      Contact
    </div>
  </div>
  <div
    class="page__content"
  >
    Contact me!
  </div>
</div>
`;
